src/parse/function.rs

```
@@ -0,0 +1,97 @@
+
+// TODO:
+// - Add documentation
+// - Add more tests
+// - Standardized user facing API.
+
+use super::utils::*;
+use super::*;
+use std::iter::Peekable;
+
+#[derive(Debug)]
+struct Function {
+    visibility: Visibility,
+    is_async: bool,
+    is_unsafe: bool,
+
+    name: String,
+    generics: Option<Generics>,
+
+    // TODO:
+    // args: Vec<FnArg>,
+    // where_cl : WhereClause,
+    // ret_ty : ReturnType,
+    // body: FnBody,
+
+    #[allow(dead_code)]
+    rest: TokenStream, // For debugging purposes
+}
+
+impl Function {
+    pub(crate) fn take(input: &mut Peekable<impl Iterator<Item = TokenTree>>) -> Self {
+        let visibility = Visibility::take(input);
+        let is_async = consume_ident_if(input, "async").is_some();
+        let is_unsafe = consume_ident_if(input, "unsafe").is_some();
+
+        // Ignore everything until `fn` keyword
+        let _ = input.skip_while(|tt| tt.to_string() != "fn").next();
+
+        let name = consume_ident(input)
+            .expect("Expected: `fn <name>`")
+            .to_string();
+
+        let generics = Generics::try_take(input).unwrap();
```

This should also return an error